src/ensemble/src/main/scala/EnsembleByKey.scala

```diff
@@ -0,0 +1,187 @@
+import org.apache.spark.ml.linalg.SQLDataTypes._
+import org.apache.spark.ml.linalg.{DenseVector, Vector, Vectors}
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.{DefaultParamsReadable, Identifiable}
+import org.apache.spark.sql.expressions.{MutableAggregationBuffer, UserDefinedAggregateFunction}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Column, DataFrame, Dataset, Row}
+
+import scala.collection.mutable
+
+object EnsembleByKey extends DefaultParamsReadable[EnsembleByKey]
+
+class EnsembleByKey(val uid: String) extends Transformer with MMLParams {
+  def this() = this(Identifiable.randomUID("EnsembleByKey"))
+
+  val keys = new StringArrayParam(this, "keys", "keys to group by")
+
+  def getKeys: Array[String] = $(keys)
+
+  def setKeys(arr: Array[String]): this.type = set(keys, arr)
+
+  def setKeys(arr: String*): this.type = set(keys, arr.toArray)
+
+  def setKey(value: String): this.type = set(keys, Array(value))
+
+  val cols = new StringArrayParam(this, "cols", "Cols to ensemble")
+
+  def getCols: Array[String] = $(cols)
+
+  def setCols(arr: Array[String]): this.type = set(cols, arr)
+
+  def setCols(arr: String*): this.type = set(cols, arr.toArray)
+
+  def setCol(value: String): this.type = set(cols, Array(value))
+
+  val colNames = new StringArrayParam(this, "colNames", "Names of the result of each col")
+
+  def getColNames: Array[String] = $(colNames)
+
+  def setColNames(arr: Array[String]): this.type = set(colNames, arr)
+
+  def setColNames(arr: String*): this.type = set(colNames, arr.toArray)
+
+  def setColName(value: String): this.type = set(colNames, Array(value))
+
+  val allowedStrategies = Set("mean")
+  val strategy = StringParam(this, "strategy", "How to ensemble the scores ex:" +
+    " mean", { x: String => allowedStrategies(x) })
+
+  def getStrategy: String = $(strategy)
+
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  setDefault(strategy -> "mean")
+
+  val collapseGroup = BooleanParam(this, "collapseGroup", "whether to collapse all items in group to one entry")
+
+  def getCollapseGroup: Boolean = $(collapseGroup)
+
+  def setCollapseGroup(value: Boolean): this.type = set(collapseGroup, value)
+
+  setDefault(collapseGroup -> true)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+
+    if (get(colNames).isEmpty) {
+      setDefault(colNames -> getCols.map(name => s"avg($name)"))
+    }
+
+    val strategyToFloatFunction = Map(
+      "mean" -> { (x: String, y: String) => mean(x).alias(y) }
+    )
+
+    val newCols = getCols.zip(getColNames).map { case (inColName, outColName) =>
+      dataset.schema(inColName).dataType match {
+        case _: DoubleType =>
+          strategyToFloatFunction(getStrategy)(inColName, outColName)
+        case _: FloatType =>
+          strategyToFloatFunction(getStrategy)(inColName, outColName)
+        case v if v == VectorType && getStrategy == "mean" =>
```

Comment:
Please create a ``Map`` similar to ``strategyToFloatFunction`` that can store the UDF so you don't have to special-case them here.